analyzers/src/SonarAnalyzer.Common/Rules/ExpectedExceptionAttributeShouldNotBeUsedBase.cs

```diff
@@ -33,16 +34,29 @@ public abstract class ExpectedExceptionAttributeShouldNotBeUsedBase<TSyntaxKind>
         protected ExpectedExceptionAttributeShouldNotBeUsedBase() : base(DiagnosticId) { }
 
         protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+            context.RegisterCompilationStartAction(c =>
             {
-                if (HasMultiLineBody(c.Node)
-                    && !AssertInCatchFinallyBlock(c.Node)
-                    && c.SemanticModel.GetDeclaredSymbol(c.Node) is { } methodSymbol
-                    && methodSymbol.GetAttributes(UnitTestHelper.KnownExpectedExceptionAttributes).FirstOrDefault() is { } attribute)
+                if (!ContainExpectedExceptionType(c.Compilation))
                 {
-                    c.ReportIssue(Rule, attribute.ApplicationSyntaxReference.GetSyntax());
+                    return;
                 }
-            },
-            Language.SyntaxKind.MethodDeclarations);
+
+                c.RegisterNodeAction(Language.GeneratedCodeRecognizer, cc =>
+                    {
+                        if (HasExpectedExceptionAttribute(cc.Node)
+                            && HasMultiLineBody(cc.Node)
+                            && !AssertInCatchFinallyBlock(cc.Node)
+                            && cc.SemanticModel.GetDeclaredSymbol(cc.Node) is { } methodSymbol
+                            && methodSymbol.GetAttributes(UnitTestHelper.KnownExpectedExceptionAttributes).FirstOrDefault() is { } attribute)
```

Comment:
Rename the `HasExpectedExceptionAttribute` to `FindHasExpectedExceptionAttribute` (or something similar) and make it return the attribute syntax (as `SyntaxNode` type) when found, `null` otherwise. Then you can skip the use of the semantic model, which is a nice improvement.

Reply:
We still need to use the semantic model to ensure the ExpectedException type other we might have some FPs. 
Could be OK I guess.

Reply:
True, but how likely is an FP like that? We check that it's a unit test project (MsTest or lower version NUnit) and it has a custom attribute called `ExpectedException` that is used to decorate methods? I'd say the chances for this are astronomical.
But this was more of a suggestion rather than a blocking comment. I'll leave it up to you.